js/src/main/scala/fm/common/rich/RichElement.scala

```diff
@@ -45,4 +74,86 @@ final class RichElement(val elem: Element) extends AnyVal {
   def addClass(className: String): Unit = if (!elem.classList.contains(className)) elem.classList.add(className)
   
   def removeClass(className: String): Unit = elem.classList.remove(className)
+
+  /** Toggle between two classes, e.g. ({{{icon.toggleClass("glyphicon-eye-close", "glyphicon-eye-open")}}})*/
+  def toggleClass(valueA: String, valueB: String): Unit = {
+    if (elem.hasClass(valueA)) {
+      elem.removeClass(valueA)
+      elem.addClass(valueB)
+    } else {
+      elem.removeClass(valueB)
+      elem.addClass(valueA)
+    }
+  }
+
+  def getParentElement: Option[Element] = {
+    if (elem.parentNode.isNull) return None
+
+    elem.parentNode match {
+      case e: Element => Some(e)
+      case _ => None
+    }
+  }
+
+  def hasParentElement: Boolean = getParentElement.isDefined
+
+  def hasChildElements: Boolean = {
+    if (!hasChild) return false
+
+    elem.childNodes.exists{ _.isInstanceOf[Element] }
+  }
+
+  def childElements: Seq[Element] = {
+    if (!hasChild) return Nil
+
+    elem.childNodes.collect {
+      case e: Element => e
+    }
+  }
+
+  /** Find the closest ancestor of the current element (or the current element itself) that matches a class AND selector. */
+  def closest[A <: Element : ClassTag](selector: String): A = closestImpl(elem, selector)
+
+  @tailrec
+  private def closestImpl[A <: Element : ClassTag](element: Element, selector: String): A = {
+    if (classTag[A].runtimeClass.isInstance(element) && element.matches(selector)) element.asInstanceOf[A]
+    else if (element.hasParent) closestImpl[A](element.getParentElement.get, selector)
+    else throw new NoSuchElementException(s"No Matching Parent for ${classTag[A]}")
+  }
+
+  /** Find the closest ancestor or ancestor's child of the current element that matches a class AND selector. */
```

Comment:
I'm not sure what this means.  Can you add examples to the documentation?

The `closestAncestorOrAncestorChildImpl` code checks `elem` and then `elem`'s child elements and then goes up to the parent?  That doesn't seem to match up with this comment.

Reply:
`/** Find the closest ancestor or ancestors' children (of the current element) that matches a class AND selector. (e.g. closestAncestorOrAncestorChild("input[type='password']") will find the closest password input field that was defined before the current element) */`

Is that better, or still obtuse as a name?  `closest` only checks ancestors, it doesn't check their children.

Reply:
```
html
  body
    div.one
      div.two <-- 1. calling closestAncestorOrAncestorChild[Div](".three") will return the child div.three or the sibling?
        div.three
      input   <-- 2. calling closestAncestorOrAncestorChild[HTMLInputElement]("input") will return this same element?
      div.two <-- 3. calling closestAncestorOrAncestorChild[Div]("div") will return this element?
      div.three
```

 - 1 - Returns the child?
 - 2 - Returns itself?
 - 3 - Returns itself or its parent?